plugins/transport-nio/src/main/java/org/elasticsearch/http/nio/NioHttpResponse.java

```diff
@@ -19,19 +19,100 @@
 
 package org.elasticsearch.http.nio;
 
-import io.netty.handler.codec.http.FullHttpResponse;
+import io.netty.handler.codec.http.DefaultFullHttpResponse;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import org.elasticsearch.common.bytes.BytesReference;
 import org.elasticsearch.http.HttpPipelinedMessage;
+import org.elasticsearch.http.HttpResponse;
+import org.elasticsearch.rest.RestStatus;
 
-public class NioHttpResponse extends HttpPipelinedMessage {
+import java.util.Collections;
+import java.util.EnumMap;
+import java.util.Map;
 
-    private final FullHttpResponse response;
+public class NioHttpResponse extends DefaultFullHttpResponse implements HttpResponse, HttpPipelinedMessage {
 
-    public NioHttpResponse(int sequence, FullHttpResponse response) {
-        super(sequence);
-        this.response = response;
+    private final int sequence;
+    private final NioHttpRequest request;
+
+    NioHttpResponse(NioHttpRequest request, RestStatus status, BytesReference content) {
+        super(request.nettyRequest().protocolVersion(), getStatus(status), ByteBufUtils.toByteBuf(content));
+        this.sequence = request.sequence();
+        this.request = request;
+    }
+
+    @Override
+    public void addHeader(String name, String value) {
+        headers().add(name, value);
+    }
+
+    @Override
+    public boolean containsHeader(String name) {
+        return headers().contains(name);
+    }
+
+    @Override
+    public int getSequence() {
+        return sequence;
+    }
+
+    private static Map<RestStatus, HttpResponseStatus> MAP;
+
+    public NioHttpRequest getRequest() {
+        return request;
+    }
+
+    static {
+        EnumMap<RestStatus, HttpResponseStatus> map = new EnumMap<>(RestStatus.class);
```

Comment:
can we somehow share this ? I saw it in multiple places now. maybe in `HttpResponse`?

Reply:
Yes using the status code (int). I'll do that in a followup though.